### DISCUSSION TITLE
FastApi and nicegui

### DISCUSSION THREAD

**Discussion by markospy** [zauberzeug/nicegui#2223] [Q&A] [answered]
Hi, is there any repository in the community that exemplifies the integration of nicegui with an API created in FastApi? 

Or a little tutorial on how to do it.

Thank you in advance.

**Comment by hasansezertasan** [comment]
What do you mean by integration? Using an API as a third party server or using NiceGUI inside a FastAPI Application?

Please check this out before:

https://github.com/zauberzeug/nicegui/tree/e926909f594bbd192ea54a4c5924e90ce01d0870/examples/fastapi

**Reply by markospy** [reply]
I mean to use nicegui inside an API created by me in FastApi. That is, using nicegui for the front-end of my application in FastApi. The example of the repository is not very clear to me.

**Reply by rodja** [reply]
Have you run the example? Do you have any specific questions? Where is the example confusing? We want to improve and are happy about any feedback.

**Reply by hasansezertasan** [reply, marked as answer]
I believe this is quite simple:

```python
from fastapi import FastAPI
from nicegui import app as nicegui_app, ui

app = FastAPI()


@app.get('/')
def read_root():
    return {'Hello': 'World'}


# Register a page with a custom path
@ui.page('/show')
def show():
    # Your UI code goes here
    ui.label('Hello, FastAPI!')
    # Some bindings
    ui.dark_mode().bind_value(nicegui_app.storage.user, 'dark_mode')
    ui.checkbox('dark mode').bind_value(nicegui_app.storage.user, 'dark_mode')


# Integrate with your FastAPI Application
ui.run_with(
    app=app,
    storage_secret='pick your private secret here',
)
```

`storage_secret` is based on `SessionMiddleware`...

**Reply by markospy** [reply]
> Have you run the example? Do you have any specific questions? Where is the example confusing? We want to improve and are happy about any feedback. 

Thank you very much for the prompt reply. I was trying the sample code and I understood everything the first time.

**Reply by markospy** [reply]
> I believe this is quite simple:
> 
> ```python
> from fastapi import FastAPI
> from nicegui import app as nicegui_app, ui
> 
> app = FastAPI()
> 
> 
> @app.get('/')
> def read_root():
>     return {'Hello': 'World'}
> 
> 
> # Register a page with a custom path
> @ui.page('/show')
> def show():
>     # Your UI code goes here
>     ui.label('Hello, FastAPI!')
>     # Some bindings
>     ui.dark_mode().bind_value(nicegui_app.storage.user, 'dark_mode')
>     ui.checkbox('dark mode').bind_value(nicegui_app.storage.user, 'dark_mode')
> 
> 
> # Integrate with your FastAPI Application
> ui.run_with(
>     app=app,
>     storage_secret='pick your private secret here',
> )
> ```
> 
> `storage_secret` is based on `SessionMiddleware`...

Thank you very much

**Reply by hasansezertasan** [reply]
You're welcome. If that resolves, can you mark this as "answer"?

**Reply by markospy** [reply]
> You're welcome. If that resolves, can you mark this as "answer"?

I am creating my application to keep track of the blood pressure and heart rate of a group of patients. The first thing I did was to create an API with FastAPI and now I'm creating the graphical interface with nicegui. The experience has been good in general. However, I have not been able to take advantage of the end points created with FastAPI in Nicegui. I have had to re-declare functions to make queries to the database and I feel that I am not being efficient, so why then create the endpoints with fastapi.

**Reply by shato20** [reply]
> I believe this is quite simple:
> 
> ```python
> from fastapi import FastAPI
> from nicegui import app as nicegui_app, ui
> 
> app = FastAPI()
> 
> 
> @app.get('/')
> def read_root():
>     return {'Hello': 'World'}
> 
> 
> # Register a page with a custom path
> @ui.page('/show')
> def show():
>     # Your UI code goes here
>     ui.label('Hello, FastAPI!')
>     # Some bindings
>     ui.dark_mode().bind_value(nicegui_app.storage.user, 'dark_mode')
>     ui.checkbox('dark mode').bind_value(nicegui_app.storage.user, 'dark_mode')
> 
> 
> # Integrate with your FastAPI Application
> ui.run_with(
>     app=app,
>     storage_secret='pick your private secret here',
> )
> ```
> 
> `storage_secret` is based on `SessionMiddleware`...

Can somebody please help me with this example? i am trying to integrate fastAPI app with nice GUI. i wanted to copy this example to have a starting point. the example runs just fine. but doesn't start any server like i usually have with ''ui.run()"
am i missing something?

**Reply by falkoschindler** [reply]
@shato20 When attaching NiceGUI to a FastAPI app using `ui.run_with`, you need to start the server yourself, either like we show in our [FastAPI example](https://github.com/zauberzeug/nicegui/tree/main/examples/fastapi):
https://github.com/zauberzeug/nicegui/blob/30334ddece490397ce8cf53a475b230fddd24838/examples/fastapi/start.sh#L13
or following the FastAPI documentation: https://fastapi.tiangolo.com/#run-it

**Comment by markospy** [comment]
I think the answer to my doubt is in this thread: https://github.com/zauberzeug/nicegui/discussions/1440 . I will try the solution suggested by rodja in a while and let you know if it has helped me.
